### PR TITLE
add bindaddress option

### PR DESCRIPTION
**Pull request by jhunt-steds**
#### Pull Request (PR) description
Add the `bindaddress` configuration option to chrony.conf.

This will allow users to specify an IP address on an interface on which chronyd should listen.

#### This Pull Request (PR) fixes the following issues
n/a

### CODE REVIEW

**Comment by jhunt-steds** [comment]
I realize the integration tests are failing for CentOS. I believe they are actually failing currently on your existing master branch as well (despite the passing badge); when I made a trivial change to README.md in my fork and had it run the test suite, it failed with the exact same errors this branch does. You can see these results in the details linked from [https://github.com/jhunt-steds/puppet-chrony/pull/2](url) . 

In case this is useful information, I ran the tests locally as well, using the instructions for Beaker and Docker at [https://github.com/voxpupuli/puppet-chrony/blob/679302f6ff7f13a196e26c23492a08259158922c/.github/CONTRIBUTING.md#unit-tests-in-docker](url) . I ran into the issue that the `--pidfile-workaround` option is presumably meant to solve (with systemd complaining it would not start chronyd because of a symlink issue with the pidfile) when I used the latest CentOS 7 container, but the tests all passed when I used a 7.6 container.

If I'm way off on this or I've done something incorrectly I beg your patience; this is the first time I've done something like this. I would be happy to provide any more information I can or make any needed changes. Thanks!

**Comment by jasonknudsen** [comment]
+1

File: manifests/init.pp [hunk 1]
```diff
@@ -187,6 +189,7 @@
 # @param dumpdir
 #   Directory to store measurement history in on exit.
 class chrony (
+  Optional[String] $bindaddress                                    = undef,
```
Comment (smortex) [inline manifests/init.pp:192]: Does it make sense to use `Stdlib::IP::Address` here?

Also, does the service allow binding to multiple IP addresses? If so, an array of IP addresses may be more suitable (defaulting to an empty array).

Comment (jasonknudsen) [reply manifests/init.pp:192]: bindaddress only seems to be able to take one argument
`chronyd[10873]: Fatal error : Too many arguments for bindaddress directive at line 12 in file /etc/chrony.conf`

Agree with changing the type to Stdlib::IP::Address. @jhunt-steds are you able to update your PR?

Comment (jhunt-steds) [reply manifests/init.pp:192]: Thanks very much to both of you for the comments and suggestions; I'll work on updating my PR at my first opportunity, today or tomorrow (end of fiscal at my org means I'm dealing with a bunch of Red Hat licensing stuff at the moment :-\ ).

Comment (jasonknudsen) [reply manifests/init.pp]: Oh my bad, it looks like you can set multiple bindaddress addresses but it needs to be multi-line like the bindcmdaddress parameter. Should definitely replicate that logic

**Comment by jhunt-steds** [comment]
I've updated my PR to use `Stdlib::IP::Address` per the comments above; the spec tests (`bundle exec rake test`) pass locally.

**Review by smortex — APPROVED** [review]

File: manifests/init.pp (hunk 1 above)
Comment (jhunt-steds) [reply manifests/init.pp]: Aha, I will look into that. Will definitely be tomorrow though.

By the way, what's the etiquette of resolving a conversation? Is that for the person who started it to do? Is it something people bother with at all?

Comment (jasonknudsen) [reply manifests/init.pp]: Believe its generally accepted that the author resolves the thread

Comment (jhunt-steds) [reply manifests/init.pp]: That makes sense, thanks.

I've updated it to use an array to allow all possible use cases as suggested. The [chrony.conf man page for the latest version](https://chrony.tuxfamily.org/doc/4.1/chrony.conf.html) says `bindaddress` can take at most one IPv4 and one IPv6 address; I tested this out on a throwaway server with one `bindaddress` directive on each line and chronyd does honor it. Anyway, I look forward to any more feedback you might have. Thanks for everything!

Comment (jasonknudsen) [reply manifests/init.pp]: I'm not 100% certain but I think you might have to revert the type, to match:
Array[String] $bindcmdaddress                                    = ['127.0.0.1', '::1'],

Comment (jasonknudsen) [reply manifests/init.pp]: Otherwise 💯  looks good to me!

Comment (jhunt-steds) [reply manifests/init.pp]: Do you mean I need to change the type of `bindaddress` from `Array[Stdlib::IP:Address]` back to `Array[String]` to match the type of `bindcmdaddress`? I understand that might be desirable for consistency's sake, but it passes a `bundle exec rake test` in its current form, and `Stdlib::IP::Address` accepts both IPv4 and IPv6 addresses, so there's no strictly technical reason I can see to go back to `String`. I think the tighter type is a good idea even if it's not quite the same as some things that were done previously, but if that consistency is valued more highly I am willing to switch it.

Comment (jasonknudsen) [reply manifests/init.pp]: Cool, okay it passes the tests! 👍

**Review by smortex — APPROVED** [review]
Looks great!

**Comment by smortex** [comment]
@jhunt-steds CI should be fixed in master.  Can you please rebase on top of it and force push to update this PR?

**Comment by jhunt-steds** [comment]
@smortex I hope I've done the right thing here--I don't have much experience with the `rebase` subcommand.

**Comment by smortex** [comment]
Weird, your commits appear twice and other commits are listed by GitHub… meh…

Let's fix this :-)

1. Ensure `master` is up-to-date on your machine:
```
git checkout master
git pull
git checkout add-bindaddress-config-opt
```

2. Rebase interactively (`-i`):
```
git rebase -i master
```

This will show up your editor with all commits between `master` and your current branch, I see all your commits twice:

```
pick c93e335 add bindaddress option
pick 863c97a change bindaddress type to Stdlib::IP::Address
pick 2168ec3 convert bindaddress to an array
pick 4b09bf5 add bindaddress option
pick 11ef42c change bindaddress type to Stdlib::IP::Address
pick ad414f2 convert bindaddress to an array
```

Just remove the first 3 lines or the last 3 lines, save and quit.  Git will remove these commits from history and the rebase should succeed!  If so, then proceed: if you `git push -f` this PR should only show your 3 commits :tada:

**Comment by jhunt-steds** [comment]
I followed your instructions exactly. Thanks for the step-by-step guide!

**Comment by smortex** [comment]
Thanks!

**Comment by jhunt-steds** [comment]
Thank _you_!

**Pull request merged**